<a href="https://colab.research.google.com/github/Alhaziq16/Group-Project-Pai/blob/main/Modelling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Modelling fahim

In [6]:
!unzip -q /content/cat_split_dataset.zip
!ls /content/cat_split_dataset

test  train  val


In [5]:
!ls /content

cat_split_dataset.zip  drive  sample_data


In [9]:
import tensorflow as tf

print("TensorFlow version:", tf.__version__)

TensorFlow version: 2.20.0


In [10]:
IMG_SIZE = (224, 224)
BATCH_SIZE = 32

train_dataset = tf.keras.utils.image_dataset_from_directory(
    "/content/cat_split_dataset/train",
    image_size=IMG_SIZE,
    batch_size=BATCH_SIZE
)

val_dataset = tf.keras.utils.image_dataset_from_directory(
    "/content/cat_split_dataset/val",
    image_size=IMG_SIZE,
    batch_size=BATCH_SIZE
)

test_dataset = tf.keras.utils.image_dataset_from_directory(
    "/content/cat_split_dataset/test",
    image_size=IMG_SIZE,
    batch_size=BATCH_SIZE,
    shuffle=False
)

Found 13029 files belonging to 10 classes.
Found 2790 files belonging to 10 classes.
Found 2803 files belonging to 10 classes.


In [17]:
print(train_dataset.class_names)

['Abyssinian', 'American Bobtail', 'American Shorthair', 'Bengal', 'Birman', 'Bombay', 'British Shorthair', 'Maine Coon', 'Ragdoll', 'Siamese']


In [11]:
print("GPU:", tf.config.list_physical_devices('GPU'))
#kena change runtime kalu dok lamo

GPU: [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


In [12]:
#part ResNet50

from tensorflow.keras.applications import ResNet50
from tensorflow.keras import layers, models

base_model = ResNet50(
    weights='imagenet',
    include_top=False,
    input_shape=(224, 224, 3)
)

base_model.trainable = False

model = models.Sequential([
    base_model,
    layers.GlobalAveragePooling2D(),
    layers.Dense(128, activation='relu'),
    layers.Dropout(0.3),
    layers.Dense(10, activation='softmax')
])

model.summary()

94765736/94765736 ━━━━━━━━━━━━━━━━━━━━ 7s 0us/step


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ resnet50 (Functional)           │ (None, 7, 7, 2048)     │    23,587,712 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ global_average_pooling2d        │ (None, 2048)           │             0 │
│ (GlobalAveragePooling2D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       262,272 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 23,851,274 (90.99 MB)

 Trainable params: 263,562 (1.01 MB)

 Non-trainable params: 23,587,712 (89.98 MB)

In [13]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [14]:
history = model.fit(
    train_dataset,
    validation_data=val_dataset,
    epochs=5
)

Epoch 1/5
408/408 ━━━━━━━━━━━━━━━━━━━━ 73s 143ms/step - accuracy: 0.5118 - loss: 1.4392 - val_accuracy: 0.5749 - val_loss: 1.2133
Epoch 2/5
408/408 ━━━━━━━━━━━━━━━━━━━━ 61s 113ms/step - accuracy: 0.5776 - loss: 1.2234 - val_accuracy: 0.6004 - val_loss: 1.1419
Epoch 3/5
408/408 ━━━━━━━━━━━━━━━━━━━━ 44s 109ms/step - accuracy: 0.6035 - loss: 1.1441 - val_accuracy: 0.6133 - val_loss: 1.1015
Epoch 4/5
408/408 ━━━━━━━━━━━━━━━━━━━━ 86s 119ms/step - accuracy: 0.6258 - loss: 1.0860 - val_accuracy: 0.6168 - val_loss: 1.1011
Epoch 5/5
408/408 ━━━━━━━━━━━━━━━━━━━━ 47s 114ms/step - accuracy: 0.6402 - loss: 1.0408 - val_accuracy: 0.6204 - val_loss: 1.0813


In [15]:
test_loss, test_acc = model.evaluate(test_dataset)

print("ResNet50 Test Accuracy:", test_acc)

88/88 ━━━━━━━━━━━━━━━━━━━━ 13s 145ms/step - accuracy: 0.6322 - loss: 1.0847
ResNet50 Test Accuracy: 0.6321797966957092


ResNet50 (5 epochs)
Validation Accuracy: 62.04%
Test Accuracy: 63.22%
Test Loss: 1.0847

In [16]:
history_2 = model.fit(
    train_dataset,
    validation_data=val_dataset,
    initial_epoch=5,
    epochs=50
)

#buat balance hok lgi 45 sbb dah ada 5 takdi

Epoch 6/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 46s 113ms/step - accuracy: 0.6541 - loss: 0.9969 - val_accuracy: 0.6133 - val_loss: 1.1123
Epoch 7/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 44s 108ms/step - accuracy: 0.6598 - loss: 0.9608 - val_accuracy: 0.6290 - val_loss: 1.0661
Epoch 8/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 45s 110ms/step - accuracy: 0.6690 - loss: 0.9326 - val_accuracy: 0.6419 - val_loss: 1.0705
Epoch 9/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 45s 110ms/step - accuracy: 0.6818 - loss: 0.8910 - val_accuracy: 0.6233 - val_loss: 1.1051
Epoch 10/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 84s 114ms/step - accuracy: 0.6928 - loss: 0.8685 - val_accuracy: 0.6240 - val_loss: 1.0936
Epoch 11/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 84s 118ms/step - accuracy: 0.6993 - loss: 0.8336 - val_accuracy: 0.6280 - val_loss: 1.1059
Epoch 12/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 82s 119ms/step - accuracy: 0.7067 - loss: 0.8156 - val_accuracy: 0.6369 - val_loss: 1.1149
Epoch 13/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 44s 109ms/step - accuracy: 0.7090 - los

In [17]:
test_loss, test_acc = model.evaluate(test_dataset)

print("Final ResNet50 Test Accuracy:", test_acc)
print("Final ResNet50 Test Loss:", test_loss)

88/88 ━━━━━━━━━━━━━━━━━━━━ 8s 86ms/step - accuracy: 0.6211 - loss: 2.1062
Final ResNet50 Test Accuracy: 0.6211202144622803
Final ResNet50 Test Loss: 2.106229305267334


| Metric              |     Result |
| ------------------- | ---------: |
| Training Accuracy   | **85.55%** |
| Validation Accuracy | **60.68%** |
| Validation Loss     | **2.0404** |


final test performance utk resNet50 accurancy: 62.11
final restNet50 test loss 2.1062

#DenseNet21

In [19]:
from tensorflow.keras.applications import DenseNet121
from tensorflow.keras import layers, models

base_model = DenseNet121(
    weights='imagenet',
    include_top=False,
    input_shape=(224, 224, 3)
)

base_model.trainable = False

densenet_model = models.Sequential([
    base_model,
    layers.GlobalAveragePooling2D(),
    layers.Dense(128, activation='relu'),
    layers.Dropout(0.3),
    layers.Dense(10, activation='softmax')
])

densenet_model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

densenet_model.summary()

29084464/29084464 ━━━━━━━━━━━━━━━━━━━━ 3s 0us/step


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ densenet121 (Functional)        │ (None, 7, 7, 1024)     │     7,037,504 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ global_average_pooling2d_1      │ (None, 1024)           │             0 │
│ (GlobalAveragePooling2D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 128)            │       131,200 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 7,169,994 (27.35 MB)

 Trainable params: 132,490 (517.54 KB)

 Non-trainable params: 7,037,504 (26.85 MB)

In [20]:
densenet_history = densenet_model.fit(
    train_dataset,
    validation_data=val_dataset,
    epochs=50
)

Epoch 1/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 113s 201ms/step - accuracy: 0.2709 - loss: 2.2492 - val_accuracy: 0.3294 - val_loss: 1.8493
Epoch 2/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 84s 101ms/step - accuracy: 0.3261 - loss: 1.8805 - val_accuracy: 0.3548 - val_loss: 1.8110
Epoch 3/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 40s 99ms/step - accuracy: 0.3478 - loss: 1.8270 - val_accuracy: 0.3401 - val_loss: 1.7607
Epoch 4/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 41s 100ms/step - accuracy: 0.3504 - loss: 1.7953 - val_accuracy: 0.3778 - val_loss: 1.7174
Epoch 5/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 41s 100ms/step - accuracy: 0.3640 - loss: 1.7749 - val_accuracy: 0.3928 - val_loss: 1.7236
Epoch 6/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 41s 100ms/step - accuracy: 0.3748 - loss: 1.7454 - val_accuracy: 0.3910 - val_loss: 1.6837
Epoch 7/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 41s 100ms/step - accuracy: 0.3714 - loss: 1.7368 - val_accuracy: 0.3964 - val_loss: 1.7241
Epoch 8/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 44s 107ms/step - accuracy: 0.3869 - loss: 1

In [21]:
test_loss, test_acc = densenet_model.evaluate(test_dataset)

print("DenseNet121 Test Accuracy:", test_acc)
print("DenseNet121 Test Loss:", test_loss)

88/88 ━━━━━━━━━━━━━━━━━━━━ 20s 226ms/step - accuracy: 0.4224 - loss: 1.6657
DenseNet121 Test Accuracy: 0.42240455746650696
DenseNet121 Test Loss: 1.6657053232192993


| Model       | Epochs | Test Accuracy |  Test Loss |
| ----------- | -----: | ------------: | ---------: |
| DenseNet121 |     50 |    **42.24%** | **1.6657** |

by using DenseNet121

#MOBILENETV3

In [22]:
from tensorflow.keras.applications import MobileNetV3Large
from tensorflow.keras import layers, models

base_model = MobileNetV3Large(
    weights='imagenet',
    include_top=False,
    input_shape=(224, 224, 3)
)

base_model.trainable = False

mobilenet_model = models.Sequential([
    base_model,
    layers.GlobalAveragePooling2D(),
    layers.Dense(128, activation='relu'),
    layers.Dropout(0.3),
    layers.Dense(10, activation='softmax')
])

mobilenet_model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

mobilenet_model.summary()

12683000/12683000 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step


Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ MobileNetV3Large (Functional)   │ (None, 7, 7, 960)      │     2,996,352 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ global_average_pooling2d_2      │ (None, 960)            │             0 │
│ (GlobalAveragePooling2D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 128)            │       123,008 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,120,650 (11.90 MB)

 Trainable params: 124,298 (485.54 KB)

 Non-trainable params: 2,996,352 (11.43 MB)

In [23]:
mobilenet_history = mobilenet_model.fit(
    train_dataset,
    validation_data=val_dataset,
    epochs=50
)


Epoch 1/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 81s 144ms/step - accuracy: 0.5753 - loss: 1.2550 - val_accuracy: 0.6462 - val_loss: 1.0354
Epoch 2/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 30s 72ms/step - accuracy: 0.6534 - loss: 1.0199 - val_accuracy: 0.6616 - val_loss: 0.9802
Epoch 3/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 40s 70ms/step - accuracy: 0.6744 - loss: 0.9408 - val_accuracy: 0.6785 - val_loss: 0.9529
Epoch 4/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 28s 69ms/step - accuracy: 0.7017 - loss: 0.8736 - val_accuracy: 0.6778 - val_loss: 0.9421
Epoch 5/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 43s 75ms/step - accuracy: 0.7144 - loss: 0.8233 - val_accuracy: 0.6763 - val_loss: 0.9464
Epoch 6/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 31s 76ms/step - accuracy: 0.7331 - loss: 0.7759 - val_accuracy: 0.6828 - val_loss: 0.9516
Epoch 7/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 31s 76ms/step - accuracy: 0.7445 - loss: 0.7353 - val_accuracy: 0.6674 - val_loss: 0.9847
Epoch 8/50
408/408 ━━━━━━━━━━━━━━━━━━━━ 38s 69ms/step - accuracy: 0.7549 - loss: 0.6966 -

In [24]:
test_loss, test_acc = mobilenet_model.evaluate(test_dataset)

print("mobilenet_model Test Accuracy:", test_acc)
print("mobilenet_model Test Loss:", test_loss)

88/88 ━━━━━━━━━━━━━━━━━━━━ 19s 216ms/step - accuracy: 0.6721 - loss: 2.0551
mobilenet_model Test Accuracy: 0.6721370220184326
mobilenet_model Test Loss: 2.055098533630371


Conclusion comparison:

| Model       | Epochs | Test Accuracy | Test Loss |
| ----------- | -----: | ------------: | --------: |
| ResNet50    |     50 |        62.11% |    2.1062 |
| DenseNet121 |     50 |        42.24% |    1.6657 |
| MobileNetV3 |     50 |        67.21% |    2.0551 |

so MobileNetv3 paling tinggi in term of test accuracy

In [33]:
#calculate 33

from sklearn.metrics import average_precision_score
from tensorflow.keras.utils import to_categorical
import numpy as np

# Get true labels
y_true = []

for images, labels in test_dataset:
    y_true.extend(labels.numpy())

y_true = np.array(y_true)

# Convert to one-hot encoding
y_true_onehot = to_categorical(y_true, num_classes=10)

print(y_true_onehot.shape)

(2803, 10)


In [34]:
#restnet50 average score

resnet_preds = model.predict(test_dataset)

resnet_map = average_precision_score(
    y_true_onehot,
    resnet_preds,
    average='macro'
)

print("ResNet50 mAP:", resnet_map)

88/88 ━━━━━━━━━━━━━━━━━━━━ 21s 144ms/step
ResNet50 mAP: 0.4851869829367043


In [35]:
#desnet
densenet_preds = densenet_model.predict(test_dataset)

densenet_map = average_precision_score(
    y_true_onehot,
    densenet_preds,
    average='macro'
)

print("DenseNet121 mAP:", densenet_map)

88/88 ━━━━━━━━━━━━━━━━━━━━ 45s 262ms/step
DenseNet121 mAP: 0.2662865149041226


In [36]:
#mobilenet

mobilenet_preds = mobilenet_model.predict(test_dataset)

mobilenet_map = average_precision_score(
    y_true_onehot,
    mobilenet_preds,
    average='macro'
)

print("MobileNetV3 mAP:", mobilenet_map)

88/88 ━━━━━━━━━━━━━━━━━━━━ 17s 131ms/step
MobileNetV3 mAP: 0.5466836644839981


| Model       | Test Accuracy (%) |    mAP | Training Time (minutes) |
| ----------- | ----------------: | -----: | ----------------------: |
| ResNet50    |             62.32 | 0.4852 |                      48 |
| DenseNet121 |             42.24 | 0.2663 |                      35 |
| MobileNetV3 |             67.21 | 0.5467 |                      31 |

by considering two metrics